ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,128 @@
+
+open D
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let inode_size = 4096L
+
+let ( + ) = Int64.add
+
+let ( // ) = Filename.concat
+
+exception Illegal_file_path of string
+
+exception Unsupported_file_type of string
+
+exception File_exceeds_max_size_limit of string
+
+exception File_incomplete of string
+
+exception File_corrupted of string
+
+exception File_parsed_failed of string
+
+let unpack dir acc_size ifd head =
+  let assert_file_path_is_legal file_path =
+    if
+      List.exists
+        (fun s ->
+          String.equal Filename.current_dir_name s
+          || String.equal Filename.parent_dir_name s
+        )
+        (String.split_on_char '/' file_path)
+    then
+      raise (Illegal_file_path file_path)
+  in
+  let assert_file_not_exceed_max_size acc_size =
+    if Int64.compare acc_size !Xapi_globs.bundle_file_max_size > 0 then
+      raise
+        (File_exceeds_max_size_limit
+           (Printf.sprintf
+              "Current total unpacked file size: %Ld, exceeds max size limit: \
+               %Ld"
+              acc_size
+              !Xapi_globs.bundle_file_max_size
+           )
+        )
+  in
+  let file_name = head.Tar.Header.file_name in
+  debug "Unpacking tar file: %s" file_name ;
+  assert_file_path_is_legal file_name ;
+  let file_path = dir // file_name in
+  match head.Tar.Header.link_indicator with
+  | Tar.Header.Link.Normal ->
+      let file_size = head.Tar.Header.file_size in
+      let new_acc_size = acc_size + file_size + inode_size in
+      Unixext.with_file file_path [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC]
+        head.Tar.Header.file_mode (fun ofd ->
+          assert_file_not_exceed_max_size new_acc_size ;
+          let real_copied_size = Unixext.copy_file ~limit:file_size ifd ofd in
+          if real_copied_size <> file_size then
+            raise (File_incomplete "Copied size is insistent with file size")
+      ) ;
+      new_acc_size
+  | Tar.Header.Link.Directory ->
+      let new_acc_size = acc_size + dir_size + inode_size in
+      assert_file_not_exceed_max_size new_acc_size ;
+      Unixext.mkdir_rec file_path head.Tar.Header.file_mode ;
+      new_acc_size
+  | Hard
+  | Symbolic
+  | Character
+  | Block
+  | FIFO
+  | GlobalExtendedHeader
+  | PerFileExtendedHeader
+  | LongLink ->
+      raise
+        (Unsupported_file_type
+           (Tar.Header.Link.to_string head.Tar.Header.link_indicator)
+        )
+
+let unpack_tar_file ~dir ~ifd =
+  debug "Unpacked to %s" dir ;
+  Unixext.rm_rec dir ;
+  Unixext.mkdir_rec dir dir_perm ;
+  let rec unpack_all_files unpack dir acc_size =
+    let acc_size = Tar_unix.Archive.with_next_file ifd (unpack dir acc_size) in
+    unpack_all_files unpack dir acc_size
+  in
+  try unpack_all_files unpack dir 0L with
+  | Tar.Header.End_of_stream ->
+      debug "Unpacked to %s successfully" dir ;
+      ()
+  | e -> (
+      error "Failed to unpack to %s with exception: %s" dir
+        (Printexc.to_string e) ;
+      match e with
```

Comment:
This can be matched directly with the `with` in `try ... with`.

Reply:
The `dir` needs to be cleaned in these error cases.

Reply:
If we want to be serious about clean-up, we could unpack into a temporary directory and remove that to clean up. Otherwise you might have trouble knowing what to remove.